components/script/script_module.rs

```diff
@@ -230,104 +225,120 @@ impl ModuleTree {
         *self.text.borrow_mut() = module_text;
     }
 
-    pub fn get_parent_urls(&self) -> &DomRefCell<IndexSet<ServoUrl>> {
-        &self.parent_urls
+    pub fn get_incomplete_fetch_urls(&self) -> &DomRefCell<IndexSet<ServoUrl>> {
+        &self.incomplete_fetch_urls
     }
 
-    pub fn insert_parent_url(&self, parent_url: ServoUrl) {
-        self.parent_urls.borrow_mut().insert(parent_url);
+    pub fn get_descendant_urls(&self) -> &DomRefCell<IndexSet<ServoUrl>> {
+        &self.descendant_urls
     }
 
-    pub fn append_parent_urls(&self, parent_urls: IndexSet<ServoUrl>) {
-        self.parent_urls.borrow_mut().extend(parent_urls);
+    pub fn get_parent_urls(&self) -> IndexSet<ServoUrl> {
+        let parent_identities = self.parent_identities.borrow();
+
+        parent_identities
+            .iter()
+            .filter_map(|parent_identity| match parent_identity {
+                ModuleIdentity::ScriptId(_) => None,
+                ModuleIdentity::ModuleUrl(url) => Some(url.clone()),
+            })
+            .collect()
     }
 
-    pub fn get_descendant_urls(&self) -> &DomRefCell<IndexSet<ServoUrl>> {
-        &self.descendant_urls
+    pub fn insert_parent_identity(&self, parent_identity: ModuleIdentity) {
+        self.parent_identities.borrow_mut().insert(parent_identity);
     }
 
-    pub fn append_descendant_urls(&self, descendant_urls: IndexSet<ServoUrl>) {
-        self.descendant_urls.borrow_mut().extend(descendant_urls);
+    pub fn insert_incomplete_fetch_url(&self, dependency: ServoUrl) {
+        self.incomplete_fetch_urls.borrow_mut().insert(dependency);
     }
 
-    /// recursively checks if all of the transitive descendants are
-    /// in the FetchingDescendants or later status
-    fn recursive_check_descendants(
-        module_tree: &ModuleTree,
-        module_map: &HashMap<ServoUrl, Rc<ModuleTree>>,
-        discovered_urls: &mut HashSet<ServoUrl>,
-    ) -> bool {
-        discovered_urls.insert(module_tree.url.clone());
+    pub fn remove_incomplete_fetch_url(&self, dependency: ServoUrl) {
+        self.incomplete_fetch_urls.borrow_mut().remove(&dependency);
+    }
 
-        let descendant_urls = module_tree.descendant_urls.borrow();
+    // Find circular dependencies in non-recursive way
+    pub fn find_circular_dependencies(&self, global: &GlobalScope) -> IndexSet<ServoUrl> {
```

Comment:
This function is inspired by [this blogpost](https://breakingcode.wordpress.com/2013/03/11/an-example-dependency-resolution-algorithm-in-python/). Should I link it as a (documentation) comment here?

Reply:
Yes please

Reply:
Added in 17ff240! Thanks!